### PR TITLE
Add support for more ops for padding elimination

### PR DESCRIPTION
**Pull request by guyang3532**
Add support for Gelu/ReduceMean/SimplifiedLayerNormalization for padding elimination

### CODE REVIEW

File: orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc [hunk 1]
```diff
@@ -361,6 +369,30 @@ void IterateSubgraphFromNode(Graph& graph,
...
+            candidate_outputs.insert(cur);
+            axes_check = false;
+            break;
+          }
+        }
+        if (axes_check) {
+          LOG_DEBUG_INFO(logger, "PaddingElimination::ReduceMean: " + cur->Name() + " is added to subgraph.");
+          subgraph.insert(cur->MutableOutputDefs()[0]);
+          PushAllOutputNode(graph, to_visit, cur, visited);
+        }
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc:390]: for axes_check being false, we need also add cur in candidate_outputs?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 2]
```diff
@@ -5811,21 +5814,21 @@ def test_elementwise(self, input_ids):
             return output
 
         # test MatMul op for padding elimination
-        # in case 0, the shapes of inputs of MatMul are [2, seqlen] and [batch_size, seqlen, hidden_size]
-        #            this case is not support in padding elimination, so the MatMul should not be included in padding
-        #            elimination subgraph and the GatherGrad should be added before MatMul.
-        # in case 1, the shapes of inputs of MatMul are [batch_size, seqlen, hidden_size] and [hidden_size, 128]
+        # in case 0, the shapes of inputs of MatMul are [batch_size, seqlen, hidden_size] and [hidden_size, 128]
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:5817]: can you add explaination for case 0/1 for non-matmul ops?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 3]
```diff
@@ -5900,13 +5910,11 @@ def find_input_node_type(model, arg):
     gathergrad_input_optypes = [find_input_node_type(training_model, arg) for arg in gathergrad_node.input]
     if test_op == "Add" or test_op == "Mul" or test_op == "Sub":
         assert test_op in gathergrad_input_optypes
-    elif test_op == "MatMul":
+    else:
         if case == 0:
+            assert test_op in gathergrad_input_optypes
+        else:
             assert "ATen" in gathergrad_input_optypes
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:5917]: Aten will be always be in gathergrad_input_optypes? 

```suggestion
            assert "ATen" in gathergrad_input_optypes
```

**Comment by pengwa** [comment]
Add @askhade and @baijumeswani for your reference。

File: orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc [hunk 4]
```diff
@@ -361,6 +369,30 @@ void IterateSubgraphFromNode(Graph& graph,
       } else {
         candidate_outputs.insert(cur);
       }
+    } else if (graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "ReduceMean", {1, 11, 13, 18})) {
+      if (cur->InputDefs()[0]->Shape()) {
+        auto axes = cur->GetAttributes().at("axes").ints();
+        bool axes_check = (axes.size() > 0);
+        for (int64_t axis : axes) {
+          axis = axis < 0 ? axis + cur->InputDefs()[0]->Shape()->dim_size() : axis;
```
[the remark this answers is not in the document]
Comment (guyang3532) [reply orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc:377]: OK

File: orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc (hunk 1 above)
Comment (guyang3532) [reply orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc:391]: Yep! you are right! thanks, done.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 2 above)
Comment (guyang3532) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:5817]: Only Elementwise ops and 'MatMul', 'ReduceMean' have multiple cases and have added comments for all them.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 3 above)
Comment (guyang3532) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:5922]: Yep, only 'MatMul' case 1 and 'ReduceMean' case 1 will go into this else, and "Aten" should be in gathergrad_input_optypes.

**Review by pengwa — APPROVED** [review]

**Pull request merged**